README
NAME
    Jifty::Plugin::OpenID - Provides OpenID authentication for your jifty
    app

DESCRIPTION
    Provides OpenID authentication for your app

USAGE
  Config
    please provide "OpenIDSecret" in your etc/config.yml , the "OpenIDUA" is
    optional , OpenID Plugin will use LWPx::ParanoidAgent by default.

        --- 
        application:
            OpenIDSecret: 1234
            OpenIDUA: LWP::UserAgent

    or you can set "OpenIDUserAgent" environment var in command-line:

        OpenIDUserAgent=LWP::UserAgent bin/jifty server

    if you are using LWPx::ParanoidAgent as your openid agent. you will need
    to provide "JIFTY_OPENID_WHITELIST_HOST" for your own OpenID server.

        export JIFTY_OPENID_WHITELIST_HOST=123.123.123.123

  User Model
    Create your user model , and let it uses
    Jifty::Plugin::OpenID::Mixin::Model::User to mixin "openid" column. and
    a "name" method.

        use TestApp::Record schema {

            column email =>
                type is 'varchar';

        };
        use Jifty::Plugin::OpenID::Mixin::Model::User;

        sub name {
            my $self = shift;
            return $self->email;
        }

    Note: you might need to declare a "name" method. because the OpenID
    CreateOpenIDUser action and SkeletonApp needs current_user->username to
    show welcome message and success message , which calls
    "brief_description" method. See Jifty::Record for "brief_description"
    method.

  View
    OpenID plugin provides AuthenticateOpenID Action. so that you can render
    an AuthenticateOpenID in your template:

        form {
            my $openid = new_action( class   => 'AuthenticateOpenID',
                                    moniker => 'authenticateopenid' );
            render_action( $openid );
        };

    this action renders a form which provides openid url field. and you will
    need to provide a submit button in your form.

        form {
            my $openid = new_action( class   => 'AuthenticateOpenID',
                                    moniker => 'authenticateopenid' );

            # ....

            render_action( $openid );
            outs_raw(
                Jifty->web->return(
                    to     => '/openid_verify_done',
                    label  => _("Login with OpenID"),
                    submit => $openid
                ));
        };

    the "to" field is for verified user to redirect to. so that you will
    need to implement a template called "/openid_verify_done":

        template '/openid_verify_done' => page {
            h1 { "Done" };
        };

  Attribute Exchange
    You can retrieve informations from remote profile on authentication
    server with OpenID Attribute Exchange service extension.

    Set in your config.yml

        - OpenID:
           ax_param: openid.ns.ax=http://openid.net/srv/ax/1.0&openid.ax.mode=fetch_request&openid.ax.type.email=http://axschema.org/contact/email&openid.ax.type.firstname=http://axschema.org/namePerson/first&openid.ax.type.lastname=http://axschema.org/namePerson/last&openid.ax.required=firstname,lastname,email
           ax_values: value.email,value.firstname,value.lastname
           ax_mapping: "{ 'email': 'value.email', 'name': 'value.firstname value.lastname' }"

    this parameters are usuable for all OpenID endpoints supporting
    Attribute Exchange extension. They can be overriden in your application.
    Watch and/or override "openid/wayf" template from
    Jifty::Plugin::OpenID::View.

    Or you can use in your view "show('openid/wayf','/url_return_to');".

   ax_param
    is the url send to authentication server. It defines namespace, mode,
    attributes types and requested attributes.

    hints : MyOpenID use schema.openid.net schemas instead of axschema.org,
    Google provides lastname and firstname, Yahoo only fullname

   ax_values
    keys of attributes values read from authentication server response.

   ax_mapping
    mapping of recieve values with our applicalication fields in json
    format.

AUTHORS
    Alex Vandiver, Cornelius <cornelius.howl {at} gmail.com >, Yves Agostini

LICENSE
    Copyright 2005-2010 Best Practical Solutions, LLC.

    This program is free software and may be modified and distributed under
    the same terms as Perl itself.



Changes
Revision history for Perl module Jifty::Plugin::OpenID

1.01 Mon, 11 Jan 2010 10:59:19 +0100
    - add Attribute Exchange management
    - bump Net::OpenID::Consumer dep

1.00 Wed, 10 Jun 2009 11:55:15 +0200
    - bump version number to avoid conflict with debian
    - add minimal load test
    - temporary suspend tests with Test::OpenID::Server

0.10 Tue, 09 Jun 2009 22:28:10 +0200
    - original version for CPAN


t/TestApp/lib/TestApp/View.pm
package TestApp::View;
use strict;
use warnings;
use Jifty::View::Declare -base;
use Jifty::View::Declare::Helpers;


template '/' => page {
    with ( name => 'openid-form' ),
    form {
        my $openid = new_action( class   => 'AuthenticateOpenID',
                                moniker => 'authenticateopenid' );
        div { { class is 'openid'};
            div { { id is 'openid-login' };
                render_action( $openid );
                outs_raw(
                    Jifty->web->return(
                        to     => '/openid_verify_done',
                        label  => _("Login with OpenID"),
                        submit => $openid
                    ));
            };
        };
    };
    show 'openid/wayf';
};


template '/openid_verify_done' => page {
    h1 { "Done" };
};




1;


lib/Jifty/Plugin/OpenID/View.pm
package Jifty::Plugin::OpenID::View;
use strict;
use warnings;
use Jifty::View::Declare -base;

=head1 NAME

Jifty::Plugin::OpenID::View - Login and create pages for OpenID plugin

=head1 DESCRIPTION

The view class for L<Jifty::Plugin::OpenID>.  Provides login and create pages.

=cut

template 'openid/login' => page {
    { title is _("Login with your OpenID") }
    my ($action, $next) = get('action', 'next');

    div {
        unless ( Jifty->web->current_user->id ) {
            div {
                attr { id => 'openid-login' };
                outs(
                    p {
                        em {
                            _(  qq{If you have a Livejournal or other OpenID account, you don\'t even need to sign up. Just log in.}
                            );
                        }
                    }
                );
                Jifty->web->form->start( call => $next );
                render_action($action);
                form_submit(
                    label  => _("Go for it!"),
                    submit => $action
                );
                Jifty->web->form->end;
            };
        }
        else {
            outs( _("You already logged in.") );
        }
    }
};

template 'openid/create' => page {
    title is _('Set your username');
    my ( $action, $next ) = get( 'action', 'next' );

    p {
        outs(
            _(  'We need you to set a username or quickly check the one associated with your OpenID. Your username is what other people will see when you ask questions or make suggestions'
            )
        );
    };
    p {
        outs(
            _(  'If the username provided conflicts with an existing username or contains invalid characters, you will have to give us a new one.'
            )
        );
    };
    Jifty->web->form->start( call => $next , name => 'openid-user-create' );
    my $openidSP = Jifty->web->session->get('ax_mapping');
    if ($openidSP) {
        foreach my $param (keys %$openidSP) {
            # keep get to use validation
            render_param($action, $param, default_value => get($param) );
        }
    }
    else {
        render_action($action);
        };
    form_submit( label => _('Continue'), submit => $action );
    Jifty->web->form->end;
};

# optionnal fragment to add direct links to Google, Yahoo,
# MyOpenID login

template 'openid/wayf' => sub {
    my ( $self, $return_to ) = @_;
    div { attr { class => ''; };
        form {
            my $google = new_action( class => 'AuthenticateOpenID', moniker => 'authenticateopenid' );
            render_param($google, 'openid', render_as => 'hidden', default_value => 'www.google.com/accounts/o8/id');
            render_param($google, 'ax_param', render_as => 'hidden', default_value => "openid.ns.ax=http://openid.net/srv/ax/1.0&openid.ax.mode=fetch_request&openid.ax.type.email=http://axschema.org/contact/email&openid.ax.type.firstname=http://axschema.org/namePerson/first&openid.ax.type.lastname=http://axschema.org/namePerson/last&openid.ax.required=firstname,lastname,email");
            render_param($google, 'ax_mapping', render_as => 'hidden', default_value => "{ 'email': 'value.email', 'name': 'value.firstname value.lastname' }");
            render_param($google, 'ax_values', render_as => 'hidden', default_value => "value.email,value.firstname,value.lastname" );
            render_param($google,'return_to', render_as => 'hidden', default_value => '/openid/verify_and_login');
            img { src is '/static/oidimg/FriendConnect.gif'; };
            outs_raw(
                Jifty->web->return(
                as_link => 1,
                to => $return_to,
                label => _("Sign in with your Google Account"),
                submit => $google
                ));
        };
        form {
            my $yahoo = new_action( class => 'AuthenticateOpenID', moniker => 'authenticateopenid' );
            render_param($yahoo, 'openid', render_as => 'hidden', default_value => 'me.yahoo.com');
            render_param($yahoo, 'ax_param', render_as => 'hidden', default_value => "openid.ns.ax=http://openid.net/srv/ax/1.0&openid.ax.mode=fetch_request&openid.ax.type.email=http://axschema.org/contact/email&openid.ax.type.fullname=http://axschema.org/namePerson&openid.ax.required=fullname,email");
            render_param($yahoo, 'ax_mapping', render_as => 'hidden', default_value => "{ 'email': 'value.email', 'name': 'value.fullname' }");
            render_param($yahoo, 'ax_values', render_as => 'hidden', default_value => "value.email,value.fullname" );
            render_param($yahoo,'return_to', render_as => 'hidden', default_value => '/openid/verify_and_login');
            img { src is '/static/oidimg/yfavicon.gif'; };
            outs_raw(
                Jifty->web->return(
                as_link => 1,
                to => $return_to,
                label => _("Sign in with your Yahoo account"),
                submit => $yahoo
                ));
        };
        form {
            my $myoid = new_action( class => 'AuthenticateOpenID', moniker => 'authenticateopenid' );
            render_param($myoid, 'openid', render_as => 'hidden', default_value => 'www.myopenid.com');
            render_param($myoid, 'ax_param', render_as => 'hidden', default_value => "openid.ns.ax=http://openid.net/srv/ax/1.0&openid.ax.mode=fetch_request&openid.ax.type.email=http://schema.openid.net/contact/email&openid.ax.type.nickname=http://schema.openid.net/namePerson/friendly&openid.ax.required=nickname,email");
            render_param($myoid, 'ax_mapping', render_as => 'hidden', default_value => "{ 'email': 'value.email.1', 'name': 'value.nickname.1' }");
            render_param($myoid, 'ax_values', render_as => 'hidden', default_value => "value.email.1,value.nickname.1" );
            render_param($myoid,'return_to', render_as => 'hidden', default_value => '/openid/verify_and_login');
            img { src is '/static/oidimg/myopenid.png'; };
            outs_raw(
                Jifty->web->return(
                as_link => 1,
                to => $return_to,
                label => _("Sign in with your MyOpenID Account"),
                submit => $myoid
                ));
        };
    };
};

1;


lib/Jifty/Plugin/OpenID/Dispatcher.pm
use strict;
use warnings;

package Jifty::Plugin::OpenID::Dispatcher;
use Jifty::Dispatcher -base;

=head1 NAME

Jifty::Plugin::OpenID::Dispatcher - Dispatcher for OpenID plugin

=head1 DESCRIPTION

Dispatcher for L<Jifty::Plugin::OpenID>.  Handles a lot of the work.

=cut

before qr'^/(?:openid/link)' => run {
    tangent('/openid/login') unless (Jifty->web->current_user->id)
};

before qr'^/openid/login' => run {
    Jifty->api->allow('AuthenticateOpenID');
    set action => Jifty->web->new_action(
        class   => 'AuthenticateOpenID',
        moniker => 'authenticateopenid'
    );
};

before qr'^/openid/verify' => run {
    Jifty->api->allow('VerifyOpenID');
    Jifty->web->request->add_action(
        class   => 'VerifyOpenID',
        moniker => 'verifyopenid'
    );
};

on 'openid/verify_and_link' => run {
    my $result = Jifty->web->response->result('verifyopenid');
    my $user   = Jifty->web->current_user;
    if ( defined $result and $result->success and $user->id ) {
        my $openid = $result->content('openid');
        my ( $ret, $msg ) = $user->user_object->validate_openid( $openid );

        if ( not $ret ) {
            $result->error(_("It looks like someone is already using that OpenID."));
            redirect '/openid/link';
        }
        else {
            $user->user_object->link_to_openid( $openid );
            $result->message(_("The OpenID '$openid' has been linked to your account."));
        }
    }
    redirect '/';
};

on 'openid/verify_and_login' => run {
    my $result = Jifty->web->response->result('verifyopenid');

    if ( defined $result and $result->success ) {
        my $openid = $result->content('openid');
        my $user = Jifty->app_class('CurrentUser')->new( openid => $openid );
        Jifty->log->info("User Class: $user. OpenID: $openid");

        if ( $user->id ) {
            # Set up our login message
            $result->message( _("Welcome back, ") . $user->username . "." );

            # Actually do the signin thing.
            Jifty->web->current_user($user);
            Jifty->web->session->expires( undef );
            Jifty->web->session->set_cookie;

            if(Jifty->web->request->continuation) {
                Jifty->web->request->continuation->call;
            } else {
                redirect '/';
            }
        }
        else {
            # User needs to create account still
            Jifty->web->session->set( openid => $openid );
            Jifty->log->info("got openid: $openid");
            my $Mapping = Jifty->web->session->get('ax_mapping');
            if ($Mapping) {
                # get ns, google set is own ns, usually ext1
                # this need to be provided by csr
                my $signed = get('openid.signed');
                my $ns = $1 if $signed =~ m/ns\.(\w+?),/;
                # get values
                my $AX=();
                foreach my $param ( split ',', Jifty->web->session->get('ax_values') ) {
                    $AX->{$param} = get('openid.'.$ns.'.'.$param);
                };
                # map values
                foreach my $param (keys %$Mapping ) {
                    my $val = $Mapping->{$param};
                    $val =~ s/(value\.\w+(\.\d)?)/$AX->{$1}/g;
                    $Mapping->{$param} = $val;
                };
                $Mapping->{openid} = $openid;
            };
            my $nick = get('openid.sreg.nickname');
            if ( $nick ) {
                redirect( Jifty::Web::Form::Clickable->new( url => '/openid/create', parameters => { 
                    nickname => $nick, 
                    openid => $openid } ));
            }
            elsif ($Mapping) {
                redirect( Jifty::Web::Form::Clickable->new( url => '/openid/create', parameters => $Mapping ) );
            }
            else {
                redirect( Jifty::Web::Form::Clickable->new( url => '/openid/create' ) );
            }
        }
    }
    else {
        if(Jifty->web->request->continuation) {
            Jifty->web->request->continuation->call;
        } else {
            redirect '/openid/login';
        }
    }
};

on 'openid/create' => run {
    if ( not Jifty->web->session->get('openid') ) {
        redirect '/';
    }

    set action => Jifty->web->new_action( class => 'CreateOpenIDUser', parameters => { openid => Jifty->web->session->get("openid") } );
    set 'next' => Jifty->web->request->continuation ||
                  Jifty::Continuation->new( request => Jifty::Request->new( path => "/" ) );
};

1;


lib/Jifty/Plugin/OpenID/Action/AuthenticateOpenID.pm
use warnings;
use strict;

=head1 NAME

Jifty::Plugin::OpenID::Action::AuthenticateOpenID - make OpenId authentication

=cut

package Jifty::Plugin::OpenID::Action::AuthenticateOpenID;

use base qw/Jifty::Action/;

use LWPx::ParanoidAgent;
use Net::OpenID::Consumer;
use Cache::FileCache;

=head2 arguments

Return the OpenID URL field

=cut

use Jifty::Param::Schema;
use Jifty::Action schema {
    param 'openid' =>
        label is _('OpenID URL'),
        is mandatory,
        hints is _('For example: you.livejournal.com');

    param 'ax_param' =>
        render as 'Hidden';

    param 'ax_values' =>
        render as 'Hidden';

    param 'ax_mapping' =>
        render as 'Hidden';

    param 'return_to' =>
        render as 'Hidden',
        default is '/openid/verify_and_login';
};

=head2 take_action

Creates local user if non-existant and redirects to OpenID auth URL

=cut

use Jifty::JSON qw /jsonToObj/;

sub take_action {
    my $self   = shift;
    my $openid = $self->argument_value('openid');
    my $path   = $self->argument_value('return_to');
    my $plugin = Jifty->find_plugin('Jifty::Plugin::OpenID');
    my $ax_mapping   = $self->argument_value('ax_mapping') || $plugin->ax_mapping();
    my $ax_param   = $self->argument_value('ax_param') || $plugin->ax_param();
    my $ax_values   = $self->argument_value('ax_values') || $plugin->ax_values();

    my $baseurl = Jifty->web->url;
    my $csr = Jifty::Plugin::OpenID->get_csr( required_root => $baseurl );

    my $claimed_id = $csr->claimed_identity( $openid );

    if ( not defined $claimed_id ) {
        $self->result->error(_("Invalid OpenID URL.  Please check to make sure it is correct.  (%1)",@{[$csr->err]}));
        return;
    }

    $openid = $claimed_id->claimed_url;

    my $return_to = Jifty->web->url( path => $path );
    if(Jifty->web->request->continuation) {
        $return_to .= ($return_to =~ /\?/) ? '&' : '?';
        $return_to .= "J:C=" . Jifty->web->request->continuation->id;
    }

    my $check_url = $claimed_id->check_url( 
                        return_to  => $return_to,
                        trust_root => $baseurl,
                        delayed_return => 1
                    );
    Jifty->web->session->set(ax_mapping => jsonToObj($ax_mapping));
    Jifty->web->session->set(ax_values => $ax_values);
    $ax_param = '&'.$ax_param if $ax_param !~ m/^&/;

    Jifty->web->_redirect( $check_url . '&openid.sreg.optional=nickname'.$ax_param );
    return 1; # should never get here
}

1;


lib/Jifty/Plugin/OpenID.pm
use strict;
use warnings;

package Jifty::Plugin::OpenID;
use base qw/Jifty::Plugin/;

our $VERSION = '1.01';

=head1 NAME

Jifty::Plugin::OpenID - Provides OpenID authentication for your jifty app

=head1 DESCRIPTION

Provides OpenID authentication for your app

=head1 USAGE

=head2 Config

please provide C<OpenIDSecret> in your F<etc/config.yml> , the C<OpenIDUA> is
B<optional> , OpenID Plugin will use L<LWPx::ParanoidAgent> by default.

    --- 
    application:
        OpenIDSecret: 1234
        OpenIDUA: LWP::UserAgent

or you can set C<OpenIDUserAgent> environment var in command-line:

    OpenIDUserAgent=LWP::UserAgent bin/jifty server

if you are using L<LWPx::ParanoidAgent> as your openid agent. 
you will need to provide C<JIFTY_OPENID_WHITELIST_HOST> for your own OpenID server.

    export JIFTY_OPENID_WHITELIST_HOST=123.123.123.123

=head2 User Model

Create your user model , and let it uses
L<Jifty::Plugin::OpenID::Mixin::Model::User> to mixin "openid" column.
and a C<name> method.

    use TestApp::Record schema {

        column email =>
            type is 'varchar';

    };
    use Jifty::Plugin::OpenID::Mixin::Model::User;

    sub name {
        my $self = shift;
        return $self->email;
    }

Note: you might need to declare a C<name> method. because the OpenID
CreateOpenIDUser action and SkeletonApp needs current_user->username to show
welcome message and success message , which calls C<brief_description> method.
See L<Jifty::Record> for C<brief_description> method.

=head2 View

OpenID plugin provides AuthenticateOpenID Action. so that you can render an
AuthenticateOpenID in your template:

    form {
        my $openid = new_action( class   => 'AuthenticateOpenID',
                                moniker => 'authenticateopenid' );
        render_action( $openid );
    };

this action renders a form which provides openid url field.
and you will need to provide a submit button in your form.  

    form {
        my $openid = new_action( class   => 'AuthenticateOpenID',
                                moniker => 'authenticateopenid' );

        # ....

        render_action( $openid );
        outs_raw(
            Jifty->web->return(
                to     => '/openid_verify_done',
                label  => _("Login with OpenID"),
                submit => $openid
            ));
    };

the C<to> field is for verified user to redirect to.
so that you will need to implement a template called C</openid_verify_done>:

    template '/openid_verify_done' => page {
        h1 { "Done" };
    };

=head2 Attribute Exchange

You can retrieve informations from remote profile on authentication server with
 OpenID Attribute Exchange service extension.

Set in your config.yml

    - OpenID:
       ax_param: openid.ns.ax=http://openid.net/srv/ax/1.0&openid.ax.mode=fetch_request&openid.ax.type.email=http://axschema.org/contact/email&openid.ax.type.firstname=http://axschema.org/namePerson/first&openid.ax.type.lastname=http://axschema.org/namePerson/last&openid.ax.required=firstname,lastname,email
       ax_values: value.email,value.firstname,value.lastname
       ax_mapping: "{ 'email': 'value.email', 'name': 'value.firstname value.lastname' }"

this parameters are usuable for all OpenID endpoints supporting Attribute
Exchange extension. They can be overriden in your application. Watch and/or
override C<openid/wayf> template from L<Jifty::Plugin::OpenID::View>.

Or you can use in your view C<show('openid/wayf','/url_return_to');>.


=head3 ax_param

is the url send to authentication server. It defines namespace, mode, attributes
types and requested attributes.

hints : MyOpenID use schema.openid.net schemas instead of axschema.org, Google
provides lastname and firstname, Yahoo only fullname

=head3 ax_values

keys of attributes values read from authentication server response.

=head3 ax_mapping

mapping of recieve values with our applicalication fields in json format.

=cut

__PACKAGE__->mk_accessors(qw(ax_mapping ax_values ax_param));

sub init {
    my $self = shift;
    my %opt = @_;
    my $ua_class = $self->get_ua_class;
    eval "require $ua_class";
    $self->ax_param($opt{ax_param});
    $self->ax_mapping($opt{ax_mapping});
    $self->ax_values($opt{ax_values});

    Jifty->web->add_css('openidplugin.css');
}

sub get_ua_class {
    return Jifty->config->app('OpenIDUA') 
                || $ENV{OpenIDUserAgent} 
                || 'LWPx::ParanoidAgent' ;
}

sub new_ua {
    my $class = shift;
    my $ua;
    my $ua_class = $class->get_ua_class;

    Jifty->log->info( "OpenID Plugin is using $ua_class as UserAgent" );

    if( $ua_class eq 'LWPx::ParanoidAgent' ) {
         $ua = LWPx::ParanoidAgent->new(
                        whitelisted_hosts => [ $ENV{JIFTY_OPENID_WHITELIST_HOST} ]
                     );
    }
    else {
        $ua = $ua_class->new;
    }
    return $ua;
}


sub get_csr {
    my $class = shift;
    return Net::OpenID::Consumer->new(
        ua              => $class->new_ua ,
        cache           => Cache::FileCache->new,
        args            => scalar Jifty->handler->cgi->Vars,
        consumer_secret => Jifty->config->app('OpenIDSecret'),
        @_,
    );
}

=head1 AUTHORS

Alex Vandiver, Cornelius  <cornelius.howl {at} gmail.com >, Yves Agostini

=head1 LICENSE

Copyright 2005-2010 Best Practical Solutions, LLC.

This program is free software and may be modified and distributed under the same terms as Perl itself.

=cut

1;
